File: Modules/ServerTypes/Announcement/RecurringAnnouncementForm.ts
import { StringEditor, TextAreaEditor, MultipleImageUploadEditor, BooleanEditor, LookupEditor, DateEditor, IntegerEditor, PrefixedContext } from "@serenity-is/corelib";
import { AnnouncementGeneratedEditor } from "@/HumanResource/Announcement/AnnouncementGenerated/AnnouncementGeneratedEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface RecurringAnnouncementForm {
    Name: StringEditor;
    AnnouncementContent: TextAreaEditor;
    UploadDocument: MultipleImageUploadEditor;
    BindToOccupation: BooleanEditor;
    BindToDepartment: BooleanEditor;
    BindToDivision: BooleanEditor;
    BindToJobGrade: BooleanEditor;
    BindToSection: BooleanEditor;
    OccupationListActual: LookupEditor;
    DepartmentListActual: LookupEditor;
    DivisionListActual: LookupEditor;
    JobGradeListActual: LookupEditor;
    SectionListActual: LookupEditor;
    EmployeeListActual: LookupEditor;
    All: BooleanEditor;
    StartingDateTime: DateEditor;
    RecurringTime: StringEditor;
    DaysOfWeekRecurring: BooleanEditor;
    DaysPerRecurring: BooleanEditor;
    IntervalInDays: IntegerEditor;
    Sunday: BooleanEditor;
    Monday: BooleanEditor;
    Tuesday: BooleanEditor;
    Wednesday: BooleanEditor;
    Thursday: BooleanEditor;
    Friday: BooleanEditor;
    Saturday: BooleanEditor;
    Remarks: TextAreaEditor;
    AnnouncementList: AnnouncementGeneratedEditor;
}

export class RecurringAnnouncementForm extends PrefixedContext {
    static formKey = 'Announcement.RecurringAnnouncement';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!RecurringAnnouncementForm.init)  {
            RecurringAnnouncementForm.init = true;

            var w0 = StringEditor;
            var w1 = TextAreaEditor;
            var w2 = MultipleImageUploadEditor;
            var w3 = BooleanEditor;
            var w4 = LookupEditor;
            var w5 = DateEditor;
            var w6 = IntegerEditor;
            var w7 = AnnouncementGeneratedEditor;

            initFormType(RecurringAnnouncementForm, [
                'Name', w0,
                'AnnouncementContent', w1,
                'UploadDocument', w2,
                'BindToOccupation', w3,
                'BindToDepartment', w3,
                'BindToDivision', w3,
                'BindToJobGrade', w3,
                'BindToSection', w3,
                'OccupationListActual', w4,
                'DepartmentListActual', w4,
                'DivisionListActual', w4,
                'JobGradeListActual', w4,
                'SectionListActual', w4,
                'EmployeeListActual', w4,
                'All', w3,
                'StartingDateTime', w5,
                'RecurringTime', w0,
                'DaysOfWeekRecurring', w3,
                'DaysPerRecurring', w3,
                'IntervalInDays', w6,
                'Sunday', w3,
                'Monday', w3,
                'Tuesday', w3,
                'Wednesday', w3,
                'Thursday', w3,
                'Friday', w3,
                'Saturday', w3,
                'Remarks', w1,
                'AnnouncementList', w7
            ]);
        }
    }
}

File: Modules/HumanResource/Announcement/RecurringAnnouncement/RecurringAnnouncementForm.cs
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace HRMSoftware.Announcement.Forms;

[FormScript("Announcement.RecurringAnnouncement")]
[BasedOnRow(typeof(RecurringAnnouncementRow), CheckNames = true)]
public class RecurringAnnouncementForm
{

    [Tab("Announcement Content")]
    [Category("Content")]

    public string Name { get; set; }
    [TextAreaEditor(Rows = 5)]
    public string AnnouncementContent { get; set; }
    public string UploadDocument { get; set; }



    [Tab("Settings")]

    [Category("Binds")]
    [QuarterWidth]
    public bool BindToOccupation { get; set; }
    [QuarterWidth]
    public bool BindToDepartment { get; set; }
    [QuarterWidth]
    public bool BindToDivision { get; set; }
    [QuarterWidth]
    public bool BindToJobGrade { get; set; }
    [QuarterWidth]
    public bool BindToSection { get; set; }

    [Category("Filters")]

   

    [HalfWidth]
    public List<int> OccupationListActual { get; set; }
    [HalfWidth]
    public List<int> DepartmentListActual { get; set; }
    [HalfWidth]
    public List<int> DivisionListActual { get; set; }
    [HalfWidth]
    public List<int> JobGradeListActual { get; set; }
    [HalfWidth]
    public List<int> SectionListActual { get; set; }
    [ThreeQuarterWidth]
    public List<int> EmployeeListActual { get; set; }
   
    
    [QuarterWidth]
    public int All { get; set; }



    [Tab("Intervals")]


    [HalfWidth]
    public DateTime StartingDateTime { get; set; }
    [HalfWidth]
    public string RecurringTime { get; set; }

    [QuarterWidth]
    public bool DaysOfWeekRecurring { get; set; }
    [QuarterWidth]
    public bool DaysPerRecurring { get; set; }


    [Category("Recurring Days")]
    [QuarterWidth]
    public int IntervalInDays { get; set; }


    [Category("Days of Week")]
    public bool Sunday { get; set; }
    public bool Monday { get; set; }
    public bool Tuesday { get; set; }
    public bool Wednesday { get; set; }
    public bool Thursday { get; set; }
    public bool Friday { get; set; }
    public bool Saturday { get; set; }

    [Category("Remarks")]
    [TextAreaEditor(Rows = 5)]
    public string Remarks { get; set; }



    [Tab("Announcement Generated"), Category(""), IgnoreName, LabelWidth("0"), AnnouncementGeneratedEditor,HideOnInsert]
    public List<AnnouncementGeneratedRow> AnnouncementList { get; set; }

}